File: tvOS/xCodeProj/ImmichChannelTV/ImmichChannelTV/AppConfig.swift
import Foundation

struct AppConfig: Codable, Equatable {
    var immichURL: String = ""
    var apiKey: String = ""
    var minDuration: Double = 10
    var randomBatchSize: Int = 20
    var onlyFavorites: Bool = false
    var debug: Bool = false
    var crossfadeEnabled: Bool = true
    var crossfadeDurationMs: Int = 450
    var preloadSecondsBeforeEnd: Double = 4
    var queueTargetSize: Int = 2
    var playbackQuality: String = "auto"
    var useSQLiteCache: Bool = true
    var syncOnStartup: Bool = true
    var syncPageSize: Int = 200
    var syncMaxPages: Int = 200

    var isConfigured: Bool {
        !immichURL.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var normalizedImmichBaseURL: String {
        immichURL.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    var playbackPeakBitRate: Double {
        switch playbackQuality {
        case "low":
            return 2_000_000
        case "medium":
            return 5_000_000
        case "high":
            return 10_000_000
        default:
            return 0
        }
    }

    var playbackQualityLabel: String {
        switch playbackQuality {
        case "low":
            return "low"
        case "medium":
            return "med"
        case "high":
            return "high"
        default:
            return "auto"
        }
    }
}

@MainActor
final class ConfigStore: ObservableObject {
    @Published var config: AppConfig

    private let defaultsKey = "ImmichChannelTV.Config.v1"

    init() {
        if let data = UserDefaults.standard.data(forKey: defaultsKey),
           let decoded = try? JSONDecoder().decode(AppConfig.self, from: data) {
            self.config = decoded
        } else {
            self.config = AppConfig()
        }
    }

    func save(_ newConfig: AppConfig) {
        config = newConfig
        persist()
    }

    func reset() {
        config = AppConfig()
        persist()
    }

    private func persist() {
        if let data = try? JSONEncoder().encode(config) {
            UserDefaults.standard.set(data, forKey: defaultsKey)
        }
    }
}

File: android/app/src/main/java/uk/co/bananasystems/homevideochannel/AppConfig.kt
package uk.co.bananasystems.homevideochannel

data class AppConfig(
    val immichUrl: String = "",
    val apiKey: String = "",
    val minDuration: Double = 10.0,
    val randomBatchSize: Int = 20,
    val onlyFavorites: Boolean = false,
    val debug: Boolean = false,
    val crossfadeEnabled: Boolean = true,
    val crossfadeDurationMs: Int = 450,
    val preloadSecondsBeforeEnd: Double = 4.0,
    val queueTargetSize: Int = 2,
    val playbackOrder: String = "random",
    val playbackQuality: String = "auto",
    val showDateLocationOverlay: Boolean = true,
    val useSQLiteCache: Boolean = true,
    val syncOnStartup: Boolean = true,
    val syncPageSize: Int = 200,
    val syncMaxPages: Int = 200
) {
    val isConfigured: Boolean
        get() = immichUrl.trim().isNotEmpty() && apiKey.trim().isNotEmpty()

    val normalizedImmichBaseUrl: String
        get() = immichUrl.trim().trimEnd('/')

    val playbackPeakBitrate: Long
        get() = when (playbackQuality) {
            "low" -> 2_000_000L
            "medium" -> 5_000_000L
            "high" -> 10_000_000L
            else -> 0L
        }
}
